gotrue/_async/storage.py

```diff
@@ -0,0 +1,32 @@
+from abc import ABC, abstractmethod
+from typing import Dict, Optional
+
+
+class AsyncSupportedStorage(ABC):
+    @abstractmethod
+    async def get_item(self, key: str) -> Optional[str]:
+        ...
+
+    @abstractmethod
+    async def set_item(self, key: str, value: str) -> None:
+        ...
+
+    @abstractmethod
+    async def remove_item(self, key: str) -> None:
+        ...
+
+
+class AsyncMemoryStorage(AsyncSupportedStorage):
```

Comment:
Why does this entire file exist (I know why, unasync being naive)? It does nothing asynchronously.
This should be pulled out into a top-level `storage.py` file, instead of giving it a misleading name.

Reply:
It is true that the implementation of in-memory storage does not perform any asynchronous action, but in `AsyncGoTrueClient` a storage that inherits from `AsyncSupportedStorage` must be used and therefore `AsyncMemoryStorage` must define its methods asynchronously.

Reply:
> but in AsyncGoTrueClient a storage that inherits from AsyncSupportedStorage must be used and therefore AsyncMemoryStorage must define its methods asynchronously.
Only because you've type-hinted it as such?
You should move the storages into it's own top-level `storage.py` module, that will work in both the sync and async clients. Naming something `Async` when it does nothing asynchronously is misleading for end users.

Reply:
I don't think so, the zen of Python says "Explicit is better than implicit". Each client (asynchronous and synchronous) must have its base storage class from which the custom implementations will inherit, for each case it is clear that the overridden methods must be asynchronous or synchronous depending on the case.

And yes, type annotations are something that should always be taken into account, it makes the code more robust and easier to understand as the zen says "Readability counts"

Reply:
How is being explicitly misleading better than making a single class? I am not telling you to remove the type hint either.

Instead of
```
gotrue
|- _async
   |- client.py
   |- storage.py
   |- api.py
|- _sync
   |- client.py
   |- storage.py
   |- api.py
```, where the storage.py is just a complete duplicate but with misleading names, you do
```
gotrue
|- _async
   |- client.py
   |- api.py
|- _sync
   |- client.py
   |- api.py
|- storage.py
```, where the storage contains just the synchronous storage class, that both sync and async clients use. This way you don't mislead end users into thinking that there's any "asynchronous storage" operation going on at all.

Reply:
To be clear, I don't have any issue with your base class from which custom implementations inherit. What I'm talking about is an unnecessary and misleading AsyncStorage class.

Reply:
It is not unnecessary or misleading in my opinion, an asynchronous storage is necessary for example, if you want to save the information in a Redis DB instance.